tools/system_libs.py

```diff
@@ -574,99 +579,126 @@ class Dummy(object):
+
+    # Recursively add dependencies
+    for d in lib.deps:
+      add_library(system_libs_map[d])
+
+  # Go over libraries to figure out which we must include
+  for lib in system_libs:
+    if lib.shortname in included:
+      continue
+    force_this = lib.shortname in force
+    if not force_this and only_forced:
+      continue
+    include_this = force_this or lib.shortname in include
+
+    if not include_this:
+      need_syms = set()
+      has_syms = set()
       for symbols in symbolses:
         if shared.Settings.VERBOSE:
           logging.debug('undefs: ' + str(symbols.undefs))
-        for library_symbol in library_symbols:
+        for library_symbol in lib.symbols:
           if library_symbol in symbols.undefs:
-            need.add(library_symbol)
+            need_syms.add(library_symbol)
           if library_symbol in symbols.defs:
-            has.add(library_symbol)
-      for haz in has: # remove symbols that are supplied by another of the inputs
-        if haz in need:
-          need.remove(haz)
+            has_syms.add(library_symbol)
+      for haz in has_syms:
+        if haz in need_syms:
+          # remove symbols that are supplied by another of the inputs
+          need_syms.remove(haz)
       if shared.Settings.VERBOSE:
-        logging.debug('considering %s: we need %s and have %s' % (name, str(need), str(has)))
-    if force_this or (len(need) and not only_forced):
-      # We need to build and link the library in
-      logging.debug('including %s' % name)
-
-      def do_create():
-        return create(name)
+        logging.debug('considering %s: we need %s and have %s' % (shortname, str(need_syms), str(has_syms)))
+      if not len(need_syms):
+        continue
 
-      libfile = shared.Cache.get(name, do_create, extension=suffix)
-      ret.append(libfile)
-      force = force.union(deps)
+    # We need to build and link the library in
+    add_library(lib)
 
   if shared.Settings.WASM_BACKEND:
-    ret.append(shared.Cache.get('wasm_compiler_rt.a', lambda: create_wasm_compiler_rt('wasm_compiler_rt.a'), extension='a'))
-    ret.append(shared.Cache.get('wasm_libc_rt.a', lambda: create_wasm_libc_rt('wasm_libc_rt.a'), extension='a'))
-
-  ret.sort(key=lambda x: x.endswith('.a')) # make sure to put .a files at the end.
-
-  for actual in ret:
-    if os.path.basename(actual) == 'libcxxabi.bc':
-      # libcxxabi and libcxx *static* linking is tricky. e.g. cxa_demangle.cpp disables c++
-      # exceptions, but since the string methods in the headers are *weakly* linked, then
-      # we might have exception-supporting versions of them from elsewhere, and if libcxxabi
-      # is first then it would "win", breaking exception throwing from those string
-      # header methods. To avoid that, we link libcxxabi last.
-      ret = [f for f in ret if f != actual] + [actual]
+    ret.append((shared.Cache.get('wasm_compiler_rt.a', lambda: create_wasm_compiler_rt('wasm_compiler_rt.a'), extension='a'), False))
```

Comment:
I'm not fond of the tuple-ing of this, but I can't think of a better way to do it. :/